### PR TITLE
[WIP] Send presence to joining host only

### PR DESCRIPTION
**Pull request by Half-Shot**
Attempting to fix https://github.com/matrix-org/synapse/issues/3962

(I am very new to the federation side of things, please be gentle)

As far as I can tell, there isn't a way to force `send_presence` to only send to a subset of hosts in the room and so I'm introducing one. This PR is very WIP and isn't ready for a formal review.

**EDIT** in terms of how effective this is:

Before

After

### CODE REVIEW

**Comment by Half-Shot** [comment]
This would have worked had it not been for `send_queue`. I might need to pick some brains on the best way to engineer this thing over replication.

File: synapse/federation/transaction_queue.py [hunk 1]
```diff
@@ -279,19 +279,31 @@ def _send_pdu(self, pdu, destinations):
...
-    def send_presence(self, states):
+    def send_presence(self, states, servers=None):
         """Send the new presence states to the appropriate destinations.
 
         This actually queues up the presence states ready for sending and
         triggers a background task to process them and send out the transactions.
 
         Args:
             states (list(UserPresenceState))
+            servers [String] | None
```
Comment (anoadragon453) [inline synapse/federation/transaction_queue.py]:
```suggestion
            servers (list(str)|None): A list of servers to send presence to.
```

File: synapse/federation/transaction_queue.py [hunk 2]
```diff
@@ -279,19 +279,31 @@ def _send_pdu(self, pdu, destinations):
...
         """
         if not self.hs.config.use_presence:
             # No-op if presence is disabled.
             return
 
+        if servers is not None:
+            for destination in servers:
+                logger.info("Sending targetted presence to %s", destination)
```
Comment (anoadragon453) [inline synapse/federation/transaction_queue.py]:
```suggestion
                logger.info("Sending targeted presence to %s", destination)
```

**Comment by Half-Shot** [comment]
Supersedes by #4942

**Pull request closed without merging**